File: trunk/sources/lexicon.cpp
#include "lexicon.h"

Lexicon::Lexicon(const QString& path)
{
    this->filePath = path;
    this->dictionary = new QStringList();
    this->lexicon = new QStringList();

    this->load();
}

Lexicon::~Lexicon()
{
    this->dictionary->clear();
    delete this->dictionary;

    this->lexicon->clear();
    delete this->lexicon;
}

void Lexicon::load()
{
    QFile file(this->filePath);

    if (! file.open(QIODevice::ReadOnly | QIODevice::Text))
        return;

    QTextStream in(&file);
    while (! in.atEnd())
       this->dictionary->append(in.readLine());

    file.close();
}

void Lexicon::filterLexicon(QString pattern)
{
    // first clear out the existing lexicon
    this->lexicon->clear();

    // build the regexp
    QRegExp rx(QString("[%1]+").arg(pattern.toLower()));

    for (int i = 0; i < this->dictionary->size(); i++)
    {
        QString word = this->dictionary->at(i);

        if (rx.exactMatch(word))
            this->lexicon->append(word);
    }
}

QStringListIterator Lexicon::iterator()
{
    return QStringListIterator(*this->lexicon);
}

bool Lexicon::hasWord(const QString& val)
{
    // QStringList must implement binarySearch into contains; this is quite fast
    return this->lexicon->contains(val, Qt::CaseInsensitive);
}

File: csharp/Boggle/Lexicon.cs
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;

class Lexicon
{
    private string file { get; set; }
    private List<string> dictionary;
    private List<string> lexicon;

    public Lexicon(string file)
    {
        this.file = file;
        this.dictionary = new List<string>();
        this.lexicon = new List<string>();

        this.load();
    }

    private void load()
    {
        using (StreamReader reader = new StreamReader(this.file))
        {
            string line;
            while ((line = reader.ReadLine()) != null)
            {
                this.dictionary.Add(line);
            }
        }
    }

    public void filterLexicon(string pattern)
    {
        // clear the lexicon first
        this.lexicon.Clear();

        // build a regexp that will only allow certain characters
        Regex rx = new Regex("[" + pattern + "]+", RegexOptions.Compiled | RegexOptions.IgnoreCase);

        foreach (string s in this.dictionary)
            if (rx.IsMatch(s))
                this.lexicon.Add(s);
    }

    public bool hasWord(string word)
    {
        return (this.lexicon.BinarySearch(word) >= 0);
    }

    public IEnumerator<string> GetEnumerator()
    {
        return this.lexicon.GetEnumerator();
    }
}